### PR TITLE
swarm/api: url scheme bzz-hash to get hashes of swarm content (#15238)

### PR DESCRIPTION
**Pull request by janos**
Update URI to support bzz-hash scheme and handle such HTTP requests by
responding with hash of the content as a text/plain response.

More discussion on this subject is on a related PR ethersphere#141 and gist https://gist.github.com/gbalint/ce09ada0bebc38f3a8a3dd2ce6c6299f.

### CODE REVIEW

**Comment by cobordism** [comment]
What should be the returned hash of bzz-hash://somename.eth where somename.eth resolves to the hash of a manifest (Hash H1) which contains only a default entry (empty path) for some file (Hash H2)?

FWIW i think we should change the way default entries are represented in manifests. See also https://github.com/ethersphere/go-ethereum/issues/162

**Comment by gbalint** [comment]
@homotopycolimit Imho it should just return the hash of the manifest, it is a low level feature anyway. Or maybe `300 multiple choices`?

**Comment by cobordism** [comment]
there are no multiple choices really. The manifest has only a single entry.
The question thus is, does bzz-hash://XYZ return the hash of the item bzz-raw://XYZ (the manifest) or the hash of the item bzz://XYZ (the content)

bzz-hash feel like it should be low level, and so return the hash of the manifest, but then there is no way to use it to return the hash of the content (unless / until we change the way default entries are handled in manifest. see also: https://github.com/ethersphere/go-ethereum/issues/162#issuecomment-351378344)

**Comment by janos** [comment]
@homotopycolimit I think that your are right. Returning the hash of the content is mire useful and bzz-hash should return hash of the bzz:// content, not the manifest bzz-hash://. Do you agree to make such change?

**Comment by gbalint** [comment]
For the note: we agreed on the standup that we should return the hash of the default content, and not the manifest itself

**Comment by cobordism** [comment]
Just to clarify:
so that means, that if I ask for bzz-hash:// of a manifest with many entries (I uploaded a directory for example) then I get the hash of the directory's root manifest, but if I also specify that the default file in the directory is index.html then I get the content-hash for this file?

Or was the decision to only return the hash of the default content if there is no other entry in the manifest? I ask specifically, because that is the case I am trying to get rid of in my suggestion for updating how defaults work in manifests. https://gist.github.com/homotopycolimit/7b70436326535ef19028f836680a1f3c

**Comment by gbalint** [comment]
I think we should just return the hash of that content what a simple `bzz://` request would return, no matter how manifests work. Isn't this a clear definition?

**Comment by janos** [comment]
Yes @gbalint, and that is how it works with the latest commit.

**Comment by cobordism** [comment]
Now however we have strayed from the original intent for bzzh as I understand it (to extend resolving ENS)
https://github.com/ethereum/go-ethereum/issues/15238

1. I upload a directory to swarm with default entry index.html
2. I save the hash of the root manifest in ENS at theswarm.eth
3. bzzh://theswarm.eth returns the hash of index.html and not the hash stored on ENS

Update 2 (PS): I think this problem goes away if manifests don't have content saved at the empty path: https://gist.github.com/homotopycolimit/7b70436326535ef19028f836680a1f3c in which we can always differentiate between theswarm.eth (manifest) and theswarm.eth/index.hmtl (default content).
Meanwhile bzz would redirect the browser from one to the other

**Comment by gbalint** [comment]
@nagydani WDYT?
You argued that `bzzh://` should be on the same abstraction level as `bzz://`, we just can't fulfill all requirements it seems.

**Comment by lmars** [comment]
My 2 cents:

* `bzzh://example.eth` should return the hash associated with `example.eth` in ENS (i.e. the hash of the root manifest). I often want this hash, but rarely the hash of the default content.

* if I want the hash of the content stored at `bzz://example.eth/path/to/file.txt` I can get that using a list query:

```
$ curl -s http://localhost:8500/bzz:/example.eth/path/to/file.txt?list=true
{
  "entries": [
    {
      "hash": "380629fb266a3a7f1de5fb40c4c2d1ec7ec749319e2addb5eb5b8a386b067740",
      "path": "path/to/file.txt",
      "mode": 420,
      "size": 5018,
      "mod_time": "2017-09-22T23:16:30+01:00"
    }
  ]
}
```

* if I want the hash of the default entry, I can also get that with a list query and selecting the entry with path `/`:

```
$ curl -s http://localhost:8500/bzz:/example.eth/?list=true | jq '.entries[] | select(.path == "/")'
{
  "hash": "cac9f3ac2834b0525c4dbc29136e3fd0843e3a687e0ce21b03581c4c6d4b89c7",
  "path": "/",
  "mode": 420,
  "size": 17367,
  "mod_time": "2017-12-14T15:46:48Z"
}
```

**Comment by gbalint** [comment]
I like @lmars 's solution, if @nagydani can also agree, and do not mind that `bzzh://` is a lower-level request type than `bzz://`, then we can just go with this.

**Comment by janos** [comment]
I must say that, with @lmars solution to get the content hash, I agree that returning the ens hash is more useful in this case.

**Comment by lmars** [comment]
I realised that, if the list query also returned the root hash, we would not need `bzzh://` at all, for example:

```
$ curl -s http://localhost:8500/bzz:/example.eth/?list=true | jq .
{
  "root_hash": "7a563ee6a19072de8fcb2908392dc32b7b11970f8abfa304c38474acdf7b50f1",
  "common_prefixes": [
    "docs/",
    "images/"
  ],
  "entries": [
    {
      "hash": "011b4d03dd8c01f1049143cf9c4c817e4b167f1d1b83e5c6f0f10d89ba1e7bce",
      "path": "index.html",
      "contentType": "text/html; charset=utf-8",
      "mode": 420,
      "mod_time": "2017-12-15T16:53:38Z"
    }
  ]
}
```

So I can get the root hash with:

```
$ curl -s http://localhost:8500/bzz:/example.eth/?list=true | jq .root_hash
"7a563ee6a19072de8fcb2908392dc32b7b11970f8abfa304c38474acdf7b50f1"
```

This can also be made available in the browser by just having a `h2` element with the root hash in the HTML, and also have the hashes for each piece of content in the table: https://github.com/ethereum/go-ethereum/blob/master/swarm/api/http/templates.go#L41-L50

WDYT?

**Comment by gbalint** [comment]
@lmars What if it is not a `bzzh` request for a manifest?

**Comment by lmars** [comment]
@gbalint a very valid point, I clearly rushed to a conclusion there 😊 . So really `bzzh` is just a way to resolve ENS names to hashes via the HTTP gateway right?

**Comment by gbalint** [comment]
@lmars yes, it is basically that

**Comment by janos** [comment]
It was decided in a team meeting that bzz-hash should be on the abstraction level with bzz-raw, returning the hash of the manifest, not the content.

This is implemented currently on the branch and all conflict from master branch resolved.

May I ask for a formal review approval before submitting a PR to ethereum/go-ethereum master? @gbalint @lmars @homotopycolimit

**Review by gbalint — APPROVED** [review]

File: swarm/api/http/server_test.go [hunk 1]
```diff
@@ -104,6 +104,35 @@ func TestBzzrGetPath(t *testing.T) {
...
+		url := srv.URL + "/bzz-hash:/"
+		if k[:] != "" {
+			url += common.ToHex(key[0])[2:] + "/" + k[1:]
+		}
+		resp, err = http.Get(url)
+		if err != nil {
+			t.Fatalf("Request failed: %v", err)
+		}
+		defer resp.Body.Close()
+		respbody, err = ioutil.ReadAll(resp.Body)
```
Comment (lmars) [inline swarm/api/http/server_test.go:120]: check `err`

File: swarm/api/uri.go [hunk 2]
```diff
@@ -56,7 +58,8 @@ type URI struct {
 // * <scheme>://<addr>
 // * <scheme>://<addr>/<path>
 //
-// with scheme one of bzz, bzz-raw, bzz-immutable or bzz-list
+
```
Comment (lmars) [inline swarm/api/uri.go]: Drop this extra newline

File: swarm/api/http/server_test.go (hunk 1 above)
Comment (janos) [reply swarm/api/http/server_test.go:120]: Thanks, fixed.

File: swarm/api/uri.go (hunk 2 above)
Comment (janos) [reply swarm/api/uri.go]: Dropped.

**Review by lmars — APPROVED** [review]

**Pull request closed without merging**